File: onboarding/src/main/java/br/com/accenture/wallet/onboarding/service/OnboardingService.java
package br.com.accenture.wallet.onboarding.service;

import br.com.accenture.wallet.onboarding.domain.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import org.springframework.web.context.annotation.ApplicationScope;

@Service
@ApplicationScope
public class OnboardingService {
    private final Logger log = LoggerFactory.getLogger(OnboardingService.class.getName());

    private CustomerClient customerClient;

    private AccountClient accountClient;

    public OnboardingService(CustomerClient customerClient, AccountClient accountClient) {
        this.customerClient = customerClient;
        this.accountClient = accountClient;
    }

    public String create(CustomerModel model) {
        try {
            final CustomerModel customer = createCustomer(model.getName(), model.getEmail());
            createAccount(customer);
            return customer.getId();
        } catch (Exception ex) {
            log.error(ex.getMessage(), ex);
            throw new IllegalArgumentException(ex.getMessage());
        }
    }

    private CustomerModel createCustomer(final String name, String email) {
        final CustomerModel model = new CustomerModel().setName(name).setEmail(email);
        final CustomerModel customer = customerClient.createCustomer(model);
        log.info("Customer was created with ID {}", customer.getId());
        return customer;
    }

    private void createAccount(final CustomerModel customer) {
        final AccountModel model = new AccountModel()
            .withCustomer(customer.getId())
            .withType("payment");

        final AccountModel account = accountClient.createAccount(model);
        log.info("Account was created with ID {} for customer {}", account.getId(), customer.getId());
    }

}

File: onboarding/src/main/kotlin/br/com/accenture/wallet/onboarding/service/OnboardingService.kt
package br.com.accenture.wallet.onboarding.service

import br.com.accenture.wallet.onboarding.domain.AccountBean
import br.com.accenture.wallet.onboarding.domain.BalanceBean
import br.com.accenture.wallet.onboarding.domain.CustomerBean
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.beans.factory.config.ConfigurableBeanFactory.SCOPE_SINGLETON
import org.springframework.context.annotation.Scope
import org.springframework.http.HttpEntity
import org.springframework.http.HttpHeaders
import org.springframework.http.MediaType.APPLICATION_JSON
import org.springframework.stereotype.Service
import org.springframework.web.client.RestTemplate

@Service
@Scope(SCOPE_SINGLETON)
class OnboardingService {

    private val log = LoggerFactory.getLogger(OnboardingService::class.java.name)

    @Value("\${customer.service}")
    private var customerServiceUri: String = ""

    @Value("\${account.service}")
    private var accountServiceUri: String = ""

    @Value("\${balance.service}")
    private var balanceServiceUri: String = ""

    private val headers = HttpHeaders()

    private val client = RestTemplate()

    constructor() {
        headers.contentType = APPLICATION_JSON
    }

    fun create(model: CustomerBean): Boolean {
        return try {
            val customer = createCustomer(model.name, model.email)
            val account = createAccount(customer)
            createBalance(account)
            true
        } catch (ex: Exception) {
            log.error(ex.message, ex)
            false
        }
    }

    private fun createCustomer(name: String, email: String): CustomerBean {
        val model = CustomerBean("").withName(name).withEmail(email)
        val data = HttpEntity(model, headers)
        val customer = client.postForObject(customerServiceUri, data, CustomerBean::class.java)!!
        log.info("Customer was created with ID " + customer.id)
        return customer
    }

    private fun createAccount(customer: CustomerBean): AccountBean {
        val account = AccountBean(null, customer.id!!, "payment")
        val accountData = HttpEntity(account, headers)
        return client.postForObject(accountServiceUri, accountData, AccountBean::class.java)!!
    }

    private fun createBalance(account: AccountBean): BalanceBean {
        val balance = BalanceBean(account.customer, account.id!!, 10.0)
        val balanceData = HttpEntity(balance, headers)
        return client.postForObject(balanceServiceUri, balanceData, BalanceBean::class.java)!!
    }

}
